A47-xds-federation.md

```diff
@@ -0,0 +1,682 @@
+the functionality in an application.
+
+#### `xds` Resolver Changes
+
+The xds resolver will be responsible for using the bootstrap information
+above to determine what Listener resource name to subscribe to for a
+given target URI.  The logic will be (pseudocode):
+
+```
+target_hostname = StripLeadingSlash(target_uri.path)
+if target_uri.authority is set:
+  authority_config = bootstrap.LookupAuthority(target_uri.authority)
+  if authority_config not found:
+    handle as invalid target URI
+  else:
+    template = authority_config.client_listener_resource_name_template
+    if template not set:
+      template = ("xdstp://" + target_uri.authority +
+                  "/envoy.config.listener.v3.Listener/%s")
+    xds_resource_name = ExpandPercentS(template,
+                                       PercentEncode(target_hostname))
+else:  # target_uri.authority not set
+  template = bootstrap.client_default_listener_resource_name_template
+  if template not set:
+    template = "%s"
+  if template starts with "xdstp:":
+    target_hostname = PercentEncode(target_hostname)
+  xds_resource_name = ExpandPercentS(template, target_hostname)
+```
+
+#### xDS-Enabled gRPC Server Changes
+
+Similarly to the xds resolver on the client side, the xDS-enabled gRPC
+server will need to be changed such that if the
+`server_listener_resource_name_template` field in the bootstrap file
+starts with `xdstp:`, then it will percent-encode the listening address
+before replacing the `%s` token in the template.
+
+#### `XdsClient` Changes
+
+This design will require a number of changes in `XdsClient`.
+
+##### Server Channel Map
+
+Currently, the `XdsClient` contains a single server channel.  This will be
+replaced with a map of server channels, keyed by the server definition
+from the bootstrap file (i.e., target URI, channel creds, and all known
+server features -- unknown features can be removed).  This key is used
+to de-dup servers, so that if multiple authorities use the same server,
+we do not create duplicate channels.
+
+We don't want to proactively connect a channel that is not actually
+being used.  Implementations can either proactively create the channel
+up-front and depend on idleness behavior, or they can lazily create the
+channel when it's actually needed.  Similarly, if a channel stops being
+used, implementations can either depend on idleness behavior to tear
+down the connections, or they can ref-count the channel and shut it down
+when it's no longer needed.  (The expected tear-down after being idle is
+O(5-15m).)
+
+##### Resource Cache
+
+The resource cache in the `XdsClient` is currently a two-level map, keyed
+first by resource type and then by resource name.  This will need to be
+changed to use a three-level map, keyed first by authority, then by
+resource type, and then by resource name.
+
+The top-level map in the cache is keyed by authority.  Resources without
+any authority will use the empty string as the key.  The value of this
```

Comment:
> Resources without any authority

The old-style resources don't have authority. And they use the top level server (same as before).

Is empty string a valid key for the bootstrap authority map?
If so, empty-string-authority can be different from no-authority.

Reply:
Good catch!  Yes, this could be a problem in the case where the top-level server list differs from the server list in the `""` entry in the `authorities` map.  In that case, we should use the top-level server list for old-style resource names, but we should use the server list in the `""` entry for new-style names with no authority.  (So the difference here is actually old-style names vs. new-style names; for a new-style name, there is no difference between empty-string authority and no authority.)

I've updated the text here to clarify.

@donnadionne We didn't handle this case properly in grpc/grpc#27570.  Can you please put together a PR to fix this?  We can probably handle it by adding a prefix to the authority that we use in the map key, so that old-style resources use authority "old:" and new-style resources use authority "xdstp:<authority_name>" (where "<authority_name>" may be empty).

Reply:
I made one more small change here: now that the authority key includes whether the resources are old-style or new-style, there's no more need to encode that difference in the lower level of the map.

Reply:
Donna fixed this in C-core in grpc/grpc#28106.